fv/health_test.go

```diff
@@ -0,0 +1,469 @@
+// +build fvtests
+
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fv_test
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"crypto/tls"
+	"io/ioutil"
+	"net"
+
+	"github.com/kelseyhightower/envconfig"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/onsi/gomega/types"
+	log "github.com/sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/rest"
+
+	"github.com/projectcalico/felix/fv/containers"
+	"github.com/projectcalico/felix/fv/utils"
+	"github.com/projectcalico/libcalico-go/lib/api"
+	"github.com/projectcalico/libcalico-go/lib/client"
+	"github.com/projectcalico/libcalico-go/lib/health"
+)
+
+type EnvConfig struct {
+	K8sVersion   string `default:"1.7.5"`
+	TyphaVersion string `default:"latest"`
+}
+
+var config EnvConfig
+
+var etcdContainer *containers.Container
+var apiServerContainer *containers.Container
+var k8sAPIEndpoint string
+var badK8sAPIEndpoint string
+var k8sCertFilename string
+var calicoClient *client.Client
+var k8sClient *kubernetes.Clientset
+
+var (
+	// This transport is based on  http.DefaultTransport, with InsecureSkipVerify set.
+	insecureTransport = &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+			DualStack: true,
+		}).DialContext,
+		MaxIdleConns:        100,
+		IdleConnTimeout:     90 * time.Second,
+		TLSHandshakeTimeout: 10 * time.Second,
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: true,
+		},
+		ExpectContinueTimeout: 1 * time.Second,
+	}
+	insecureHTTPClient = http.Client{
+		Transport: insecureTransport,
+	}
+)
+
+var _ = BeforeSuite(func() {
+	log.Info(">>> BeforeSuite <<<")
+	err := envconfig.Process("k8sfv", &config)
+	Expect(err).NotTo(HaveOccurred())
+	log.WithField("config", config).Info("Loaded config")
+
+	// Start etcd, which will back the k8s API server.
+	etcdContainer = containers.RunEtcd()
+	Expect(etcdContainer).NotTo(BeNil())
+
+	// Start the k8s API server.
+	apiServerContainer = containers.Run("apiserver",
+		"gcr.io/google_containers/hyperkube-amd64:v"+config.K8sVersion,
+		"/hyperkube", "apiserver",
+		fmt.Sprintf("--etcd-servers=http://%s:2379", etcdContainer.IP),
+		"--service-cluster-ip-range=10.101.0.0/16",
+		"-v=10",
+		"--authorization-mode=RBAC",
```

Comment:
Is there a reason you are turning RBAC on only to add a clusterrolebinding that gives anonymous full access? Why not `--authorization-mode=AlwaysAllow` right from the start?

Reply:
Yes: the clients in this test - Felix, Typha and the test code itself - all connect anonymously to the API server, because (a) they aren't running in pods in a proper Kubernetes cluster, and (b) they don't provide client TLS certificates, and (c) they don't use any of the other non-anonymous mechanisms that Kubernetes supports.  But, as of 1.6, the API server doesn't allow anonymous users with the default "AlwaysAllow" authorization mode.  So we specify the "RBAC" authorization mode instead, and create a ClusterRoleBinding that gives the "system:anonymous" user unlimited power (aka the "cluster-admin" role).

(This is noted in the k8sfv test at https://github.com/projectcalico/felix/blob/master/k8sfv/run-test#L108-L118, but I will add it here as well.)